### PR TITLE
Support role commands from any node

### PR DESCRIPTION
**Pull request by onurctirtir**
DESCRIPTION: Adds support from issuing role management commands from worker nodes

It's unlikely to get into a distributed deadlock with role commands, we
don't care much about them at the moment.
There were several attempts to reduce the chances of a deadlock but we
didn't any of them merged into main branch yet, see:
https://github.com/citusdata/citus/pull/7325
https://github.com/citusdata/citus/pull/7016
https://github.com/citusdata/citus/pull/7009

### CODE REVIEW

File: src/include/distributed/worker_transaction.h [hunk 1]
```diff
@@ -29,11 +29,22 @@ typedef enum TargetWorkerSet
...
+
 	/*
 	 * All the active primary nodes in the metadata except the coordinator
 	 */
 	NON_COORDINATOR_NODES,
 
+	/*
+	 * All the active primary nodes in the metadata except the local node
+	 */
+	REMOTE_NODES,
```
Comment (JelteF) [inline src/include/distributed/worker_transaction.h:46]: In #7203 this is called OTHER_NODES. I'm not sure which name I like best @halilozanakgul @onurctirtir what do you think?

File: src/backend/distributed/commands/role.c [hunk 2]
```diff
@@ -1181,11 +1198,16 @@ PreprocessGrantRoleStmt(Node *node, const char *queryString,
...
-	if (!EnableCreateRolePropagation || !IsCoordinator() || !ShouldPropagate())
+	if (!EnableCreateRolePropagation || !ShouldPropagate())
 	{
 		return NIL;
 	}
 
+	if (!IsCoordinator())
+	{
+		EnsureCoordinatorIsInMetadata();
+	}
```
Comment (JelteF) [inline src/backend/distributed/commands/role.c:1209]: This check seems common enough that we should probably make it a single function call. Maybe call it something like `EnsurePropagationToCoordinator()`

File: src/include/distributed/worker_transaction.h (hunk 1 above)
Comment (JelteF) [reply src/include/distributed/worker_transaction.h:46]: @thanodnl responded in private chat:
> REMOTE_NODES > OTHER_NODES

Comment (thanodnl) [reply src/include/distributed/worker_transaction.h:46]: `REMOTE_NODES` > `OTHER_NODES` imho

Comment (onurctirtir) [reply src/include/distributed/worker_transaction.h:46]: OTHER_NODES sounds good to me too

File: src/backend/distributed/commands/dependencies.c [hunk 3]
```diff
@@ -95,7 +95,7 @@ EnsureDependenciesExistOnAllNodes(const ObjectAddress *target)
 	 * either get it now, or get it in citus_add_node after this transaction finishes and
 	 * the pg_dist_object record becomes visible.
 	 */
-	List *workerNodeList = ActivePrimaryNonCoordinatorNodeList(RowShareLock);
+	List *workerNodeList = ActivePrimaryRemoteNodeList(RowShareLock);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/dependencies.c:98]: I have not read the code, but just wanted to let you know that @hanefi has a good context on the intricacies of DDL from any node: #7009, probably worth checking with him.

Comment (onurctirtir) [reply src/backend/distributed/commands/dependencies.c:98]: Yeah I'll talk to @hanefi when is back, thank you @onderkalaci :)

File: src/backend/distributed/commands/index.c [hunk 4]
```diff
@@ -180,6 +180,8 @@ PreprocessIndexStmt(Node *node, const char *createIndexCommand,
 		return NIL;
 	}
 
+	EnsureCoordinator();
```
Comment (JelteF) [inline src/backend/distributed/commands/index.c:183]: Why is this change needed?

File: src/test/regress/expected/metadata_sync_helpers.out [hunk 5]
```diff
@@ -38,7 +37,11 @@ BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;
 
 	SET application_name to 'citus_internal gpid=10000000001';
 	SELECT citus_internal_add_partition_metadata ('test'::regclass, 'h', 'col_1', 0, 's');
-ERROR:  This is an internal Citus function can only be used in a distributed transaction
+ citus_internal_add_partition_metadata
```
Comment (JelteF) [inline src/test/regress/expected/metadata_sync_helpers.out:40]: It's not clear to me why this change in behavior is needed for the current change.

File: src/backend/distributed/commands/index.c (hunk 4 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/index.c:183]: We were missing this check for index cmds, but luckily, we already had this in ExecuteDistributedDDLJob. However, as of this PR, ExecuteDistributedDDLJob doesn't have this check anymore. So effectively this is a no-op change for index cmds.

File: src/test/regress/expected/metadata_sync_helpers.out (hunk 5 above)
Comment (onurctirtir) [reply src/test/regress/expected/metadata_sync_helpers.out:40]: Because now EnsureCitusInitiatedOperation doesn't enforce `!IsCoordinator` anymore as we want to run some of those citus_internal UDFs in coordinator too.

File: src/include/distributed/worker_transaction.h (hunk 1 above)
Comment (JelteF) [reply src/include/distributed/worker_transaction.h]: As discussed over chat, lets go for REMOTE_NODES

**Review by JelteF — APPROVED** [review]
Looks good to me now, but as discussed over chat let's rename OTHER_NODES to REMOTE_NODES.

Also let's add to the PR description that we don't care about fixing the complicated deadlock avoidance issue for this PR (and reference the PRs that try to do that). Since deadlocks when creating/dropping/altering roles can only happen when commands on multiple nodes touch the same role. This seems very unlikely to be a common case, and if it happens a simple retry should be able to work around it easily. So the problem is small enough in practice that avoiding it is deemed unnecessary.

**Comment by onderkalaci** [comment]
> It's unlikely to get into a distributed deadlock with role commands, we
don't care much about them at the moment.

I agree, I also don't care much about deadlocks on role commands, that's probably not a common pattern to create roles concurrently.

I think the main problem with this PR is that now you opened the doors to add more DDL from worker nodes, following this exact pattern. So, in the long term, I'm afraid this might become a known-weird limitation of Citus, when you execute DDL from the workers, you might get deadlocks. Hope, DDL from workers won't evolve like that, but seem hard to avoid at this point.
